apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,744 @@
+machine but wasn't viable for production []. Fortunately, Ralf had seen this problem before and had a ready made solution:
+manually loading the shared library from within `scipy/special/__init__.py` so it would be available when needed:
+
+```python
+def _load_libsf_error_state():
+    """Load libsf_error_state.dll shared library on Windows
+    libsf_error_state manages shared state used by
+    ``scipy.special.seterr`` and ``scipy.special.geterr`` so that these
+    can work consistently between special functions provided by different
+    extension modules. This shared library is installed in scipy/special
+    alongside this __init__.py file. Due to lack of rpath support, Windows
+    cannot find shared libraries installed within wheels. To circumvent this,
+    we pre-load ``lib_sf_error_state.dll`` when on Windows.
+    The logic for this function was borrowed from the function ``make_init``
+    in `scipy/tools/openblas_support.py`:
+    https://github.com/scipy/scipy/blob/bb92c8014e21052e7dde67a76b28214dd1dcb94a/tools/openblas_support.py#L239-L274
+    """  # noqa: E501
+    if os.name == "nt":
+        try:
+            from ctypes import WinDLL
+            basedir = os.path.dirname(__file__)
+        except:  # noqa: E722
+            pass
+        else:
+            dll_path = os.path.join(basedir, "libsf_error_state.dll")
+            if os.path.exists(dll_path):
+                WinDLL(dll_path)
+
+
+_load_libsf_error_state()
+```
+
+One week in, all CI jobs were passing and the [PR](https://github.com/scipy/scipy/pull/20321) creating
+`libsf_error_state` was merged. We did it! We fixed the bug we'd introduced in `special.errstate` with months to go
+before the next SciPy release—or so we thought.
+
+### Breaking SciPy for MSVC builds
+
+On May 30th, SciPy maintainer and Conda-Forge core member Axel Obermeir (h-vetinari) posted a comment on his open PR,
+[scipy-feedstock/gh-277](https://github.com/conda-forge/scipy-feedstock/pull/277). The first release candidate for
+SciPy 1.14.0 had just come out, he was starting the process to add version 1.14.0 to Conda-Forge, and noticed that
+Windows builds for SciPy 1.14.0 were failing with the following error:
+
+```
+lld-link: error: undefined symbol: scipy_sf_error_get_action
+```
+
+How could this be? SciPy's own CI jobs were passing on Windows, but here, symbols from `lib_sf_error_state` were clearly
+not being found. The thing is, at that point in time there was still a key gap in SciPy's CI coverage. Although there
+were Windows builds in CI, they all used the MinGW compiler toolchain [], with no jobs using MSVC (Microsoft Visual C++)
+[]. We had run into another platform specific difference.
+
+Fortunately, Axel knew what the problem was. On Linux, Mac OS (and Windows while using MinGW), symbols from shared
+libraries are exported by default, but with MSVC they must be explicitly exported from shared libraries and explicitly
```

Comment:
I'd say "MSVC and Clang-cl" here, and only Clang-cl further down, since SciPy doesn't actually work with MSVC. We use Clang-cl on conda-forge.

Reply:
Ah, right. The MSVC CI job just builds it, and I remember that I could get SciPy to build with MSVC locally (if disabling Pythran), but it would fail at runtime due to an issue with Blas.